pyuvsim/telescope.py

```diff
@@ -89,31 +105,137 @@ def __init__(self, beam_list=None, uvb_params=None):
         self.uvb_params.update(list_uvb_params)
         self.uvb_params.update(uvb_params)    # Optional parameters for the UVBeam objects
 
-        x_orientation_list = []
-        for beam in beam_list:
-            if self.string_mode:
-                beam_use = self._str_to_obj(beam)
+        self.is_consistent = False
+        if check:
+            self.check_consistency(force=force_check)
+
+    def check_consistency(self, force: bool = False):
+        """
+        Check the consistency of all beams in the list.
+
+        This checks basic parameters of the objects for consistency, eg. the ``beam_type``.
+        It is meant to be manually called by the user.
+
+        Parameters
+        ----------
+        force
+            Whether to force the consistency check even if the object is in string
+            mode. This wil convert to to object mode, perform the check, then
+            convert back.
+
+        Raises
+        ------
+        BeamConsistencyError
+            If any beam is inconsistent with the rest of the beams.
+
+        """
+        if self.string_mode:
+            if not force:
+                warnings.warn(
+                    "Cannot check consistency of a string-mode BeamList! Set force=True"
+                    " to force consistency checking."
+                )
+                return
             else:
-                beam_use = beam
-
-            if hasattr(beam_use, "x_orientation"):
-                x_orientation_list.append(beam_use.x_orientation)
-        if (
-            len(x_orientation_list) == 0
-            or x_orientation_list.count(None) == len(x_orientation_list)
-        ):
-            self.x_orientation = None
-        elif all(x == x_orientation_list[0] for x in x_orientation_list):
-            self.x_orientation = x_orientation_list[0]
+                self.set_obj_mode(check=True)
+                self.set_str_mode()
+                return
+
+        if len(self) == 0:
+            return
+
+        def check_thing(item):
+            items = {i: getattr(b, item) for i, b in enumerate(self) if hasattr(b, item)}
+
+            if not items:
+                # If none of the beams has this item, move on.
+                return
+
+            min_index = min(items.keys())
+
+            for indx, thing in items.items():
+                if np.any(thing != items[min_index]):
+                    raise BeamConsistencyError(
```

Comment:
We should move the definition of the BeamConsistencyError class up to before this class, so it is defined before it is called.